plugin.php

```diff
@@ -93,94 +93,139 @@ function display_settings_page() {
+
+	// Do we need to do this, or can we just let them be?
+	$content = str_replace( "\u{00AD}", '&shy;', $content );
+
+	return $content;
 }
 
 /**
- * Adds soft hypens to any suggestion library words in the given string.
+ * Add soft hyphens to a chunk of text based on a suggestion library.
  *
- * @param string $value The string to add soft hyphens to.
+ * @param string $chunk The chunk of text to add soft hyphens to.
  *
- * @return string The string with soft hyphens added.
+ * @return string The chunk of text with soft hyphens added.
  */
-function add_soft_hyphens( string $value ) {
+function hyphenate_chunk( string $chunk ): string {
 	$hyphenation_suggestions = get_option( 'hp-soft-hyphenate' );
 
-	if ( $hyphenation_suggestions ) {
-		$words = explode( "\n", $hyphenation_suggestions );
+	if ( ! is_scalar( $hyphenation_suggestions ) || empty( $hyphenation_suggestions ) ) {
+		return $chunk;
+	}
 
-		foreach ( $words as $word ) {
-			$word = trim( $word );
+	$words = explode( "\n", (string) $hyphenation_suggestions );
+	$words = array_map( 'trim', $words );
 
-			if ( empty( $word ) ) {
-				continue;
-			}
+	preg_match_all( '/([^\s\p{P}]+)([\s\p{P}]*)/', $chunk, $matches, PREG_SET_ORDER );
 
-			// Get the word without hyphens.
-			$without_hyphens = str_replace( '-', '', $word );
+	$result = '';
+	foreach ( $matches as $match ) {
+		$hyphenated_match = $match[1];
 
-			// Create the soft-hyphenated version.
-			$soft_hyphenated = str_replace( '-', '&shy;', $word );
+		foreach ( $words as $word ) {
+			$hyphenated_match = hyphenate( $hyphenated_match, $word );
+		}
+
+		$result .= $hyphenated_match . $match[2];
+	}
+
+	return $result;
+}
 
-			// Create an all-caps soft-hyphenated version (&shy; needs to remain lowercase).
-			$soft_hyphenated_upper = str_replace( '-', '&shy;', strtoupper( $word ) );
+/**
+ * Add soft hyphens to a string based on a similar string containing
+ * suggestive hyphenation.
+ *
+ * Example: hyphenate( 'hyphenation', 'hyphenat-ion' );
+ * Returns: "hyphenat\u{00AD}ion"
+ *
+ * @param string $word       The string to add soft hyphens to.
+ * @param string $suggestion The string suggesting soft hyphen placement.
+ *
+ * @return string The string with soft hyphens added.
+ */
+function hyphenate( string $word, string $suggestion ): string {
+	$without_hyphens = str_replace( '-', '', $suggestion );
 
-			// Use case-insensitive regular expression to find matches.
-			$pattern = '/(' . preg_quote( $without_hyphens, '/' ) . ')/i';
+	if ( strtolower( $word ) !== strtolower( $without_hyphens ) ) {
+		return $word;
+	}
 
-			// Replace matches while preserving original case.
-			$value = preg_replace_callback(
-				$pattern,
-				function ( $matches ) use ( $soft_hyphenated, $soft_hyphenated_upper ) {
-					if ( strtoupper( $matches[0] ) === $matches[0] ) {
-						return $soft_hyphenated_upper;
-					}
+	// Create the soft-hyphenated version.
+	$soft_hyphenated = str_replace( '-', "\u{00AD}", $suggestion );
```

Comment:
We use `"\u{00AD}"` here for the soft hyphen because when rebuilding the HTML with the HTML API, `htmlspecialchars()` destroys `&shy;` and turns it into `&amp;shy;`.